mars/dataframe/expressions/arithmetic/add.py

```diff
@@ -55,18 +58,37 @@ def fill_value(self):
     def scalar_value(self):
         return self._scalar_value
 
+    @property
+    def lhs(self):
+        return self._lhs
+
+    @property
+    def rhs(self):
+        return self._rhs
+
+    def _set_inputs(self, inputs):
+        super(DataFrameAdd, self)._set_inputs(inputs)
+        if len(self._inputs) == 2:
+            self._lhs = inputs[0]
+            self._rhs = inputs[1]
+        else:
+            if isinstance(self._lhs, DATAFRAME_TYPE):
+                self._lhs = inputs[0]
+            elif np.isscalar(self._lhs):
+                self._rhs = inputs[0]
+
 
 def add(df, other, axis='columns', level=None, fill_value=None):
-    if np.isscalar(other):
-        op = DataFrameAdd(axis=axis, level=level, fill_value=fill_value, scalar_value=other)
+    if isinstance(other, DATAFRAME_TYPE) or np.isscalar(other):
+        op = DataFrameAdd(axis=axis, level=level, fill_value=fill_value, lhs=df, rhs=other)
     else:
-        op = DataFrameAdd(axis=axis, level=level, fill_value=fill_value)
+        raise NotImplementedError('Only support add with dataframe or scalar!')
```

Comment:
The `__add__` method could be a bit different, it must catch `NotImplementedError` and return `NotImplemented`.